frontend/lib/parsing/parser-error-classes-list.cpp

```diff
@@ -300,6 +302,23 @@ void ErrorDisallowedControlFlow::write(ErrorWriterBase& wr) const {
     }
   }
 
+  // We also special case return statements in special methods that don't allow
+  // returns with values
+  if (auto ret = invalidAst->toReturn()) {
+    if(blockingAst) {
+      if (auto fn = blockingAst->toFunction()) {
+        if((fn->name() == "deinit") || (fn->name() == "postinit")) {
```

Comment:
I think this chunk of logic could be better integrated with that above it; both need the `ret` assignment/condition, `blockingAst` condition, and `fn` assignment/condition, and you could just have an if/else on the condition `blockingAst->toFunction()->kind() == uast::Function::ITER`.

Reply:
I've refactored this code (pulling some of the nested ifs out into conditionals assigned to named variables). I think it's an improvement: more self-documenting. There's still some duplication in terms of creating the message but there's enough non-overlapping parts where I think trying to extract it would just it more difficult to read.